File: src/lib/work-revisions.ts
import type { Work, WorkRevision } from "./types";

function content(work: Work) {
  return { title: work.title, summary: work.summary, type: work.type, coverUrl: work.coverUrl, externalUrl: work.externalUrl, repositoryUrl: work.repositoryUrl, license: { ...work.license } };
}
export function currentWorkRevision(work: Work): WorkRevision {
  return work.revisions?.at(-1) ?? { id: `${work.id}:r1`, number: 1, recordedAt: "", ...content(work) };
}
export function ensureWorkRevision(work: Work, at: string) {
  if (!work.revisions?.length) work.revisions = [{ ...currentWorkRevision(work), recordedAt: at }];
  return work.revisions.at(-1)!;
}
export function recordWorkRevision(work: Work, at: string) {
  const previous = ensureWorkRevision(work, at);
  const next = content(work);
  const { id: _id, number: _number, recordedAt: _at, ...old } = previous;
  if (JSON.stringify(old) === JSON.stringify(next)) return previous;
  const number = previous.number + 1;
  const revision = { id: `${work.id}:r${number}`, number, recordedAt: at, ...next };
  work.revisions!.push(revision);
  return revision;
}

File: backend/internal/service/work_revisions.go
package service

import (
	"encoding/json"
	"fmt"
	"idea_platform/internal/domain"
	"reflect"
)

type workRevision struct {
	ID            string         `json:"id"`
	Number        int            `json:"number"`
	RecordedAt    string         `json:"recordedAt"`
	Title         string         `json:"title"`
	Summary       string         `json:"summary"`
	Type          string         `json:"type"`
	CoverURL      string         `json:"coverUrl"`
	ExternalURL   string         `json:"externalUrl,omitempty"`
	RepositoryURL string         `json:"repositoryUrl,omitempty"`
	License       domain.License `json:"license"`
}

func recordWorkRevision(work *domain.Work, at string) error {
	var revisions []workRevision
	if len(work.Revisions) > 0 {
		if err := json.Unmarshal(work.Revisions, &revisions); err != nil {
			return err
		}
	}
	next := workRevision{Title: work.Title, Summary: work.Summary, Type: work.Type, CoverURL: work.CoverURL, ExternalURL: work.ExternalURL, RepositoryURL: work.RepositoryURL, License: work.License}
	if len(revisions) > 0 {
		previous := revisions[len(revisions)-1]
		previous.ID = ""
		previous.Number = 0
		previous.RecordedAt = ""
		if reflect.DeepEqual(previous, next) {
			return nil
		}
	}
	next.Number = len(revisions) + 1
	next.ID = fmt.Sprintf("%s:r%d", work.ID, next.Number)
	next.RecordedAt = at
	revisions = append(revisions, next)
	raw, err := json.Marshal(revisions)
	if err != nil {
		return err
	}
	work.Revisions = raw
	return nil
}
